File: leetcode_vs/18.四数之和.py
#
# @lc app=leetcode.cn id=18 lang=python3
#
# [18] 四数之和
#

# @lc code=start


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if not nums:
            return []
        res = []
        lst = []
        n = len(nums)
        nums.sort()
        
        def dfs(cur_sum, start):
            # 终止条件
            if len(lst) == 4 and cur_sum == target:
                res.append(lst[:])
                return
            elif len(lst) >= 4:
                return
            for i in range(start, n):
                if n-start < 4-len(lst):
                    return
                if i > start and nums[i] == nums[i-1]:
                    continue

                check_sum = cur_sum+nums[i]
                # 剪枝
                if i < n-1 and check_sum+(3-len(lst))*nums[i+1] > target:
                    return
                if i < n-1 and check_sum+(3-len(lst))*nums[n-1] < target:
                    continue
                lst.append(nums[i])
                
                dfs(check_sum, i+1)
                lst.pop()
                
        dfs(0, 0)

        return res
# @lc code=end

File: 18.四数之和.java
/*
 * @lc app=leetcode.cn id=18 lang=java
 *
 * [18] 四数之和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        if (n < 1)
            return res;
        Arrays.sort(nums);
        for (int i = 0; i < n; i++) {
            // 因为不允许重复的三元组
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            for (int j = i + 1; j < n; j++) {
                // 对当前j找到一个最小的k使三个数加起来大于等于0
                if (j > i + 1 && nums[j] == nums[j - 1])
                    continue;
                for (int k = j + 1, u = n - 1; k < u; k++) {
                    if (k >j+1  && nums[k] == nums[k - 1])
                        continue;
                    while (k < u-1 && nums[i] + nums[j] + nums[k] + nums[u - 1] >= target)
                        u--;
                    if (nums[i] + nums[j] + nums[k] + nums[u] == target) {
                        ArrayList lst = new ArrayList<>();
                        lst.add(nums[i]);
                        lst.add(nums[j]);
                        lst.add(nums[k]);
                        lst.add(nums[u]);
                        res.add(lst);
                    }
                    
                    
                }
            }

        }
        return res;
    }
}
// @lc code=end
